File: array/ipo/ulises/ipo.py
import queue
class Project:
    def __init__(self, value):
        self.value = value
    
    def __lt__(self, next):
        return self.value[0] > next.value[0]
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        projects = []
        for i in range(len(profits)):
            p = (profits[i], capital[i])
            projects.append(p)
        projects.sort(key=lambda p: p[1])
        listIndex = 0
        pq = queue.PriorityQueue()
        for i in range(k):
            while listIndex < len(profits) and projects[listIndex][1] <= w:
                pq.put(Project(projects[listIndex]))
                listIndex += 1
            if  not pq.empty():
                w += pq.get().value[0]
        return w

File: array/ipo/ulises/ipo.cpp
class Project {
    public:
        int profit, capital;
};
bool compareProjectsByCapital(const Project &a, const Project &b)
{
   return a.capital < b.capital;
}
struct CompareProjects {
    bool operator()(const Project& a, const Project& b) {
        return a.profit < b.profit;
    }
};
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int size = profits.size();
        vector<Project> projects;
        for(int i = 0; i < size; i++)
        {
            Project p;
            p.profit = profits[i];
            p.capital = capital[i];
            projects.push_back(p);
        }
        sort(projects.begin(), projects.end(), compareProjectsByCapital);
        int totalCapital = w;
        priority_queue<Project, vector<Project>, CompareProjects> pq;
        int vectorIndex = 0;
        for(int i = 0; i < k; i++)
        {
            while(vectorIndex < size && projects[vectorIndex].capital <= totalCapital)
            {
                pq.push(projects[vectorIndex++]);
            }
            if(!pq.empty())
            {
                totalCapital += pq.top().profit;
                pq.pop();
            }
        }
        return totalCapital;
    }
};